File: hedera-mirror-rosetta/test/domain/token_builder.go
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

package domain

import (
	"fmt"

	"github.com/hashgraph/hedera-mirror-node/hedera-mirror-rosetta/app/interfaces"
	"github.com/hashgraph/hedera-mirror-node/hedera-mirror-rosetta/app/persistence/domain"
)

type TokenBuilder struct {
	dbClient interfaces.DbClient
	token    domain.Token
}

func (b *TokenBuilder) CreatedTimestamp(createdTimestamp int64) *TokenBuilder {
	b.token.CreatedTimestamp = createdTimestamp
	b.token.ModifiedTimestamp = createdTimestamp
	return b
}

func (b *TokenBuilder) Decimals(decimals int64) *TokenBuilder {
	b.token.Decimals = decimals
	return b
}

func (b *TokenBuilder) InitialSupply(initialSupply int64) *TokenBuilder {
	b.token.InitialSupply = initialSupply
	return b
}

func (b *TokenBuilder) ModifiedTimestamp(modifiedTimestamp int64) *TokenBuilder {
	b.token.ModifiedTimestamp = modifiedTimestamp
	return b
}

func (b *TokenBuilder) Type(tokenType string) *TokenBuilder {
	b.token.Type = tokenType
	return b
}

func (b *TokenBuilder) Persist() domain.Token {
	token := b.token
	b.dbClient.GetDb().Create(&token)
	return b.token
}

func NewTokenBuilder(dbClient interfaces.DbClient, tokenId, createdTimestamp, treasury int64) *TokenBuilder {
	token := domain.Token{
		CreatedTimestamp:  createdTimestamp,
		ModifiedTimestamp: createdTimestamp,
		Name:              fmt.Sprintf("%d_name", tokenId),
		SupplyType:        domain.TokenSupplyTypeInfinite,
		Symbol:            fmt.Sprintf("%d_symbol", tokenId),
		TokenId:           domain.MustDecodeEntityId(tokenId),
		TreasuryAccountId: domain.MustDecodeEntityId(treasury),
		Type:              domain.TokenTypeFungibleCommon,
	}
	return &TokenBuilder{
		dbClient: dbClient,
		token:    token,
	}
}

File: hedera-mirror-rest-java/src/test/java/com/hedera/mirror/restjava/spec/builder/TokenBuilder.java
/*
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.restjava.spec.builder;

import com.google.common.collect.Range;
import com.hedera.mirror.common.domain.entity.EntityId;
import com.hedera.mirror.common.domain.token.Token;
import com.hedera.mirror.common.domain.token.TokenFreezeStatusEnum;
import com.hedera.mirror.common.domain.token.TokenKycStatusEnum;
import com.hedera.mirror.common.domain.token.TokenPauseStatusEnum;
import com.hedera.mirror.common.domain.token.TokenSupplyTypeEnum;
import com.hedera.mirror.common.domain.token.TokenTypeEnum;
import com.hedera.mirror.common.util.DomainUtils;
import com.hedera.mirror.restjava.spec.model.SpecSetup;
import jakarta.inject.Named;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

@Named
class TokenBuilder extends AbstractEntityBuilder<Token.TokenBuilder<?, ?>> {

    @Override
    protected Supplier<List<Map<String, Object>>> getSpecEntitiesSupplier(SpecSetup specSetup) {
        return specSetup::tokens;
    }

    @Override
    protected Token.TokenBuilder<?, ?> getEntityBuilder() {
        return Token.builder()
                .createdTimestamp(0L)
                .decimals(1000)
                .freezeDefault(Boolean.FALSE)
                .freezeStatus(TokenFreezeStatusEnum.NOT_APPLICABLE)
                .initialSupply(1_000_000L)
                .kycStatus(TokenKycStatusEnum.NOT_APPLICABLE)
                .maxSupply(Long.MAX_VALUE)
                .metadata(DomainUtils.EMPTY_BYTE_ARRAY)
                .name("Token name")
                .pauseStatus(TokenPauseStatusEnum.NOT_APPLICABLE)
                .supplyType(TokenSupplyTypeEnum.INFINITE)
                .symbol("YBTJBOAZ")
                .tokenId(0L)
                .totalSupply(1_000_000L)
                .treasuryAccountId(EntityId.of(98L))
                .type(TokenTypeEnum.FUNGIBLE_COMMON);
    }

    @Override
    protected List<Object> getFinalEntities(Token.TokenBuilder<?, ?> builder, Map<String, Object> token) {
        var tokenEntity = builder.build();
        if (tokenEntity.getType() == TokenTypeEnum.NON_FUNGIBLE_UNIQUE) {
            builder.decimals(0);
            builder.initialSupply(0L);
        }

        if (tokenEntity.getTimestampRange() == null) {
            builder.timestampRange(Range.atLeast(tokenEntity.getCreatedTimestamp()));
        }

        return List.of(builder.build());
    }
}
